File: modules/http/src/main/scala/http/HttpSpec.scala
package http

import action.async.AsyncSerdes
import scala.concurrent.Future

sealed trait HttpVerb
object HttpVerb {
  case object Get    extends HttpVerb
  case object Post   extends HttpVerb
  case object Put    extends HttpVerb
  case object Delete extends HttpVerb
}

/**
  * @tparam K - key for the output topic
  * @tparam B - body for Http request
  */
final case class HttpRequest[K, B](key: K,
                                   verb: HttpVerb,
                                   url: String,
                                   headers: Map[String, String],
                                   body: Option[B],
                                   topicName: Option[String])

/**
  * @param resultDecoder
  * @param serdes
  * @param topicNames
  * @tparam K - key for the output topic
  * @tparam O - output returned by the Http request - also normally quite generic
  * @tparam R - final result type that ends up in output topic
  */
final case class HttpOutput[K, O, R](resultDecoder: O => Option[Either[Throwable, R]],
                                     serdes: AsyncSerdes[K, R],
                                     topicNames: List[String])

/**
  * @param actionType
  * @param requestDecoder
  * @param groupId
  * @param outputSpec
  * @tparam A - common representation type for all action commands (typically Json / GenericRecord for Avro)
  * @tparam K - key for the output topic
  * @tparam B - body for Http request
  * @tparam O - output returned by the Http request - also normally quite generic
  * @tparam R - final result type that ends up in output topic
  */
final case class HttpSpec[A, K, B, O, R](actionType: String,
                                         requestDecoder: A => Either[Throwable, HttpRequest[K, B]],
                                         httpClient: HttpRequest[K, B] => Future[O],
                                         groupId: String,
                                         outputSpec: Option[HttpOutput[K, O, R]])

File: java/simplesaga-http/src/main/java/io/simplesource/saga/action/http/HttpSpec.java
package io.simplesource.saga.action.http;

import io.simplesource.saga.action.async.Callback;
import lombok.Value;

import java.time.Duration;
import java.util.Optional;
import java.util.function.BiConsumer;

/**
  * @param <A> - common representation type for all action commands (typically Json / GenericRecord for Avro)
  * @param <K> - key for the output topic
  * @param <B> - body for Http request
  * @param <O> - output returned by the Http request - also normally quite generic
  * @param <R> - final result type that ends up in output topic
  */
@Value
public final class HttpSpec<A, K, B, O, R> {
    public final String actionType;
    public final HttpRequest.HttpRequestDecoder<A, K, B> decoder;
    public final BiConsumer<HttpRequest<K, B>, Callback<O>> asyncHttpClient;
    public final String groupId;
    public final Optional<HttpOutput<K, O, R>> outputSpec;
    public final Optional<Duration> timeout;
}
